lib/evaluate.js

```diff
@@ -89,19 +92,81 @@ export const evaluate = async ({
   })
 }
 
+/**
+ * @param {number} roundIndex
+ * @param {import('./typings').Measurement[]} measurements
+ * @param {import('./typings').RoundDetails} sparkRoundDetails
+ */
 export const runFraudDetection = async (roundIndex, measurements, sparkRoundDetails) => {
+  //
+  // 1. Filter out measurements not belonging to any valid task in this round
+  //    or missing some of the required fields like `inet_group`
+  //
   for (const m of measurements) {
+    if (!m.inet_group) {
+      m.fraudAssessment = 'NO_INET_GROUP'
+      console.log(m)
+      continue
+    }
+
     const isValidTask = sparkRoundDetails.retrievalTasks.some(t =>
       t.cid === m.cid && t.providerAddress === m.provider_address & t.protocol === m.protocol
     )
     if (!isValidTask) {
       m.fraudAssessment = 'INVALID_TASK'
-      continue
     }
+  }
+
+  //
+  // 2. Reward only one participant in each inet group
+  //
+  /** @type {Map<string, import('./typings').Measurement[]>} */
+  const taskGroups = new Map()
+  for (const m of measurements) {
+    if (m.fraudAssessment) continue
 
-    // TODO: add more fraud detections
+    const key = `${m.inet_group}::${m.cid}::${m.provider_address}`
+    let group = taskGroups.get(key)
+    if (!group) {
+      group = []
+      taskGroups.set(key, group)
+    }
 
-    m.fraudAssessment = 'OK'
+    group.push(m)
+  }
+
+  const getHash = (/** @type {import('./typings').Measurement} */ m) => {
+    globalThis.crypto.subtle.digest('SHA-256', Buffer.from(m.finished_at))
+  }
+
+  for (const groupMeasurements of taskGroups.values()) {
+    // Pick one measurement to reward and mark all others as not eligible for rewards
+    // The difficult part: how to choose a measurement randomly but also fairly, so
+    // that each measurement has the same chance of being picked for the reward?
+    // We also want the selection algorithm to be deterministic.
+    //
+    // Note that we cannot rely on participant addresses because it's super easy
+    // for node operators to use a different address for each measurement.
+    //
+    // Let's start with a simple algorithm we can later tweak:
+    // 1. Hash the `finished_at` timestamp recorded by the server
+    // 2. Pick the measurement with the lowest hash value
+    // This relies on the fact that the hash function has a random distribution.
+    // We are also making the assumption that each measurement has a distinct `finished_at` field.
```

Comment:
I guess even if `finished_at` values are redundant, the first measurement with the lowest hash will still win (measurement order is given), so the algorithm is still deterministic.

Why don't we reward the first measurement and instead the one with the lowest hash?

Reply:
> Why don't we reward the first measurement

Sorting by `finished_at` alone would incentivise a race between SPARK nodes. The fastest node on the quickest internet connection would consistently win the reward.

By hashing the value, we get a different value-to-compare every round.